### PR TITLE
Fix/wsdiscovery deadlock

### PR DESCRIPTION
**Pull request by deichmab-draeger**

### CODE REVIEW

File: sdc11073/wsdiscovery.py [hunk 1]
```diff
@@ -820,45 +831,42 @@ def removeSourceAddr(self, addr):
...
     def addMulticastMessage(self, env, addr, port, initialDelay=0):
         msg = Message(env, addr, port, Message.MULTICAST, initialDelay)
         self._logger.debug('addMulticastMessage: adding message Id %s. delay=%.2f'.format(env.getMessageId(), initialDelay))
-        self._queue.put(msg)
+        self._repeated_enqueue_msg(msg, initialDelay, MULTICAST_UDP_REPEAT, MULTICAST_UDP_MIN_DELAY,
+                                   MULTICAST_UDP_MAX_DELAY, MULTICAST_UDP_UPPER_DELAY)
+
+    def _repeated_enqueue_msg(self, msg, initial_delay_ms, repeat, min_delay_ms, max_delay_ms, upper_delay_ms):
+        next_send = time.time() + initial_delay_ms/1000.0
+        dt = (min_delay_ms + ((max_delay_ms - min_delay_ms) * random.random())) / 2000.0
```
Comment (ldeichmann) [inline sdc11073/wsdiscovery.py:845]: Not a fan of the magic value here. Also, consider using ` random.randrange(start, stop[, step])` to express a random integer within a range

File: sdc11073/wsdiscovery.py [hunk 2]
```diff
@@ -820,45 +831,42 @@ def removeSourceAddr(self, addr):
...
-                time.sleep(0.02)
+        while not self._quitSendEvent.is_set():
+            try:
+                enqueued_msg = self._send_queue.get(timeout=0.1)
+            except queue.Empty:
+                pass
             else:
-                time.sleep(0.2)
+                now = time.time()
+                if enqueued_msg.send_time > now:
```
Comment (ldeichmann) [inline sdc11073/wsdiscovery.py:861]: So what happens when a new message is added to the queue which needs to be send before this one? If this message is way way more delayed, it blocks other more urgent messages which have been added to the queue later on.

File: sdc11073/wsdiscovery.py [hunk 3]
```diff
@@ -992,6 +1011,7 @@ def schedule_stop(self):
         self._quitRecvEvent.set()
         self._quitSendEvent.set()
         self._recvThread.join(1)
+        self._qread_thread.join(1)
```
Comment (ldeichmann) [inline sdc11073/wsdiscovery.py:1014]: Needs to be added to the join method below. Also, add timeouts to the join method below

File: sdc11073/wsdiscovery.py (hunk 2 above)
Comment (deichmab-draeger) [reply sdc11073/wsdiscovery.py:861]: I think we can live with that. In the end this is not about priority, but only about ramdomness. The charme of this is its simplicity.

**Review by ldeichmann — APPROVED** [review]

**Pull request closed without merging**